ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,437 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        DPRINT("Munged string is '%wZ'\n", &MungedStringReal);
+    }
+
+    DPRINT("GUID string is '%wZ'\n", &GuidStringReal);
+    DPRINT("Reference string is '%wZ'\n", &ReferenceStringReal);
+
+    /* Store received parts if the parameters are not null */
+    if (PrefixString != NULL)
+    {
+        if (RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE))
+        {
+            PrefixString->Buffer = SymbolicLinkName->Buffer;
+            PrefixString->Length = KernelModePrefix.Length;
+            PrefixString->MaximumLength = PrefixString->Length;
+        }
+        else if (RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
+        {
+            PrefixString->Buffer = SymbolicLinkName->Buffer;
+            PrefixString->Length = UserModePrefix.Length;
+            PrefixString->MaximumLength = PrefixString->Length;
+        }
+
+        DPRINT("Prefix string is '%wZ'\n", PrefixString);
+    }
+
+    if (MungedString != NULL)
+        *MungedString = MungedStringReal;
     else
-        InstanceKeyRealP = &InstanceKeyReal;
+        RtlFreeUnicodeString(&MungedStringReal);
 
-    *GuidKeyRealP = NULL;
-    *DeviceKeyRealP = NULL;
-    *InstanceKeyRealP = NULL;
+    if (GuidString != NULL)
+        *GuidString = GuidStringReal;
 
-    RtlInitUnicodeString(&BaseKeyU, BaseKeyString);
+    if (ReferenceString != NULL)
+        *ReferenceString = ReferenceStringReal;
 
-    /* Open the DeviceClasses key */
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &BaseKeyU,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               NULL,
-                               NULL);
-    Status = ZwOpenKey(&ClassesKey,
-                       DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                       &ObjectAttributes);
+    if (ReferenceStringPresent != NULL)
+        *ReferenceStringPresent = ReferenceStringReal.Length != sizeof(WCHAR);
+
+    if (InterfaceClassGuid != NULL)
+    {
+        /* Convert GUID string into a GUID and store it also */
+        Status = RtlGUIDFromString(&GuidStringReal, InterfaceClassGuid);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlGUIDFromString() failed with status 0x%08lx\n", Status);
+            RtlFreeUnicodeString(&MungedStringReal);
+            RtlInitEmptyUnicodeString(MungedString, NULL, 0);
+            return Status;
+        }
+    }
+
+    /* We're done */
+    return STATUS_SUCCESS;
+}
+
+static
+NTSTATUS
+IopOpenOrCreateSymbolicLinkSubKeys(_Out_opt_ PHANDLE DeviceHandle,
+                                   _Out_opt_ PULONG DeviceDisposition,
+                                   _Out_opt_ PHANDLE InstanceHandle,
+                                   _Out_opt_ PULONG InstanceDisposition,
+                                   _In_ HANDLE ClassHandle,
+                                   _In_ PUNICODE_STRING SymbolicLinkName,
```

Comment:
Could be `PCUNICODE_STRING`, in case `IopSeparateSymbolicLink` doesn't modify it.

Reply:
Yes. It indeed does not modify an initial symbolic link.

Reply:
Applied the suggestion, thanks.